src/inmanta/execute/scheduler.py

```diff
@@ -66,9 +69,57 @@ class Scheduler(object):
     This class schedules statements for execution
     """
 
-    def __init__(self, track_dataflow: bool = False):
+    def __init__(self, track_dataflow: bool = False, type_hints: Optional[List["TypeHint"]] = None):
         self.track_dataflow: bool = track_dataflow
         self.types: Dict[str, Type] = {}
+        # The type hints specified in the project.yml file
+        self.type_hints: List["TypeHint"] = type_hints if type_hints else []
+        # A sub-list of self.type_hints containing the valid type hints.
+        # Cache field used by self.valid_type_hints property
+        self._valid_type_hints: Optional[List["TypeHint"]] = None
+
+    @property
+    def valid_type_hints(self) -> List["TypeHint"]:
+        if self._valid_type_hints is None:
+            self._valid_type_hints = self._get_valid_type_hints()
+        return self._valid_type_hints
+
+    def _get_valid_type_hints(self) -> List["TypeHint"]:
+        """
+        This method validates the type hints in self.type_hints, logs warning for invalid hints and
+        return a list containing only the valid type hints.
+        """
+        if not self.types:
+            raise Exception("The self.define_types() method should be called first")
+        if self.type_hints:
+            LOGGER.warning(
+                "[EXPERIMENTAL FEATURE] Using type hints defined in the project.yml file to determine list freeze order."
+            )
+        valid_type_hints = []
+        for hint in self.type_hints:
+            LOGGER.info("Loaded type hint: %s", hint)
+            for (entity_type_name, relationship_name) in hint.iterate_types():
+                if entity_type_name not in self.types:
+                    raise InvalidTypeHintException(
+                        f"A type hint was defined for {entity_type_name}, but no such type was defined"
+                    )
+                current_type: Type = self.types[entity_type_name]
+                if not isinstance(current_type, Entity):
+                    raise InvalidTypeHintException(f"A type hint was defined for non-entity type {current_type}")
+                assert isinstance(current_type, Entity)  # Make mypy happy
+                attributes_of_entity = current_type.attributes
+                if relationship_name not in attributes_of_entity:
+                    raise InvalidTypeHintException(
+                        f"A type hint was defined for {entity_type_name}.{relationship_name}, "
+                        f"but entity {entity_type_name} doesn't have an attribute {relationship_name}.",
+                    )
+                if not isinstance(attributes_of_entity[relationship_name], RelationAttribute):
```

Comment:
I would use this code to attach the type hint to the relation, to remove lookup all over the place, which should further improve performance.

code like `self.queues.has_type_hint(self.type)` would become `self.attribute.hints`